### DISCUSSION TITLE
Old user but still really frustrated

### DISCUSSION THREAD

**Discussion by pedz** [emacs-helm/helm#2492] [Q&A] [answered]
I usually use a new system enough to get it into my finger tips but that never happened with Helm.  I have fought against Helm the entire time I've used it and it is way past time for me to surrender.

In that quest, I come across some of the old code I've used (with comments) and one of them points to the "Tuhdo" guide but I always like to start with the bare system and then add to it.  I believe that has the best likelihood of me using what the original developer originally intended.  (This comes from some pretty awful experiences with Javascript which I won't go into.)

So... I find Tuhdo's guide and [this](http://tuhdo.github.io/helm-intro.html#orgheadline1) Usage but I'd like someone to sanity check that for me to make sure it is (more or less) aligned with the original design and is also up to date.  One of my main frustrations is (as an example) I do `M-x` which is tied to `helm-M-x`.  Often it gets into the state with a separate list which I can see but I can't get to.

To give a more complete description, Emacs has "yari" and "yari-helm" (via a package I load).  If I get into a fresh emacs and do `M-x yari-helm` it executes.  But now if I want to execute just `yari` I don't know how (until just now).  I guess I type `yari`, then do `C-v` which gets me into this second list I'm talking about.  Then I can use `C-n` to select `yari` and then hit return.  Is that how `helm` is intended to be used or is there a better way to accomplish this goal?  (By the way, I thought maybe typing `yari$` might work but it didn't.

It would probably do me some good for folks to tell me what that "second list" is called and how to intelligently ask questions about it (like I'm not doing very well right now).

TL; DR -- I'm really tired of knowing just one or two things about how to use Helm but not enough to use it effectively and productively.  I'm not working WITH Helm most of the time.  I'm looking for guidance on what to read and study to stop this pain.

**Comment by thierryvolpiatto** [comment]
Perry Smith ***@***.***> writes:

> So... I find Tuhdo's guide and this Usage but I'd like someone to
> sanity check that for me to make sure it is (more or less) aligned
> with the original design and is also up to date. One of my main
> frustrations is (as an example) I do M-x which is tied to
> helm-M-x. Often it gets into the state with a separate list which I
> can see but I can't get to.

Helm-M-x have a separate source for its history
(extended-command-history), you can jump to the next source with C-o and
back with M-o or with <right> and <left>.

> To give a more complete description, Emacs has "yari" and "yari-helm"
> (via a package I load). If I get into a fresh emacs and do M-x
> yari-helm it executes. But now if I want to execute just yari I don't
> know how (until just now). I guess I type yari, then do C-v which gets
> me into this second list I'm talking about.

I guess you bound C-v to helm-next-source (originally C-o).
Yes this is the way to do and it is what I do because when it is done
once I have yari in history and don't have anymore to jump to next
source, this is my preference, in the end all my favorite commands are
in history under hand and I rarely have to jump to next source.
This said some people (like you I presume) prefer having always the main
source on top, you can customize `helm-M-x-reverse-history` for this. 

> TL; DR -- I'm really tired of knowing just one or two things about how
> to use Helm but not enough to use it effectively and productively. I'm
> not working WITH Helm most of the time. I'm looking for guidance on
> what to read and study to stop this pain.

Did you try C-h m from any helm session?
Also from any helm source you can use C-h c to jump to customize with a
list of related user variables to configure.
Also you have the helm wiki which contains a lot of informations even if
it is not always updated nowadays.
If you don't find what you want you can always ask here I am happy to
help.


-- 
Thierry

**Reply by pedz** [reply]
I don't see the remapping of helm-next-source in my code.  I have tried C-h m but will try again.  I'm going to gut my `helm-setup` and start with just this:
```
(if (require 'helm nil t)
    (require 'helm-config))
    
(provide 'helm-setup)
```
and start from there.  I should make you helping me easier to you (I hope).

I'm sure I will be back.  Thank you for your patience.

**Reply by thierryvolpiatto** [reply]
You are welcome.

**Comment by pedz** [comment]
I added:
```
      (global-set-key (kbd "M-x") 'helm-M-x)
      (global-set-key (kbd "M-y") 'helm-show-kill-ring)
      (global-set-key (kbd "C-x b") 'helm-mini)
      (global-set-key (kbd "C-x C-f") 'helm-find-files)
```
and was playing with helm-find-files.  I noticed this text **Hit1 Expand Candidate, Hit2 or (C-u) Find file** and started to try and figure out what is trying to tell me.  I did C-h m and found this text:
```
*** ‘C-j’ on a filename expands to that filename in the Helm buffer

Second hit displays the buffer filename.
Third hit kills the buffer filename.
Note: ‘C-u C-j’ displays the buffer directly.
```
Still not understanding, I kept exploring and looking at the code and I finally understand this tiny little piece.

If I have `.con` typed so far and the helm completion window is highlighting `.config` I can hit C-j to "complete" the name -- which you call "expand the filename".  I kept trying additional C-j but I was selecting a directory.  Finally reading the code, I got the idea to try it while selecting a regular file.  The second C-j opens the filename for me to look at ... what macOS might call "preview" and the next C-j deletes the buffer and restores the screen to what it was before.  EXCEPTIONAL!!! well worth the effort to figure out.  I'm wondering if rewording it might be worth the effort.  e.g. I didn't know what you meant by "Hit1" and "Hit2".  It could be "First C-j" and "Second C-j", ...  Also "displays the buffer filename" (I think... 50/50 chance I'm wrong) would be a string that is the path to the file the buffer is displaying -- although which buffer at this point isn't clear.  To me, something like "displays the file" would be clearer to me.

One a rather extreme tangent:  Have you considered hooking up the mouse?  For example if I do `grep` and get a list of hits, I can take my mouse and click on the highlighted text of the hit and Emacs will open the file at that point.  I can effectively select the hit with my mouse as well as selecting it via keystrokes.

I've hooked up things like this long ago.  I believe Emacs calls them buttons.  And they can have tool tips that pop up.  So, for example, in the list of Actions, make each of those selectable via the mouse.  In the example we were talking about before with helm-M-x that has a second source, hovering your mouse over a candidate could pop up a tool tip of what keystrokes the user could use to select that particular item and clicking it would select the item as well.

It might be, the way helm seems to be structured, that 80% of what I'm talking about could be done by enhancing some of the lower level key functions.

**Reply by thierryvolpiatto** [reply]
Sorry if documentation is not clear, I do my best but my English is not very good.

For the mouse it is available in all helm sources when `helm-allow-mouse` is non nil, by default it is disabled because it is confusing for new users when they click outside the helm buffer and and don't have anymore control on helm because they are outside minibuffer.
The help echo is displaying only mouse bindings, for all bindings you have a section with Commands in the org buffer when you hit C-h m.

**Comment by thierryvolpiatto** [comment, marked as answer]
Here a screenshot of helm-find-files, but couldn't catch the help echo and the mouse face :-(

![Capture d’écran_2022-04-02_16-40-49](https://user-images.githubusercontent.com/1533939/161388603-cad6f567-d875-495f-bec2-fc4eed3ef308.png)

**Comment by pedz** [comment]
> On Apr 2, 2022, at 09:43, Thierry Volpiatto ***@***.***> wrote:
> 
> Sorry if documentation is not clear, I do my best but my English is not very good.

I hope I don’t sound harsh, unfriendly or ungrateful for all your work.
> For the mouse it is available in all helm sources when helm-allow-mouse is non nil, by default it is disabled because it is confusing for new users when they click outside the helm buffer and and don't have anymore control on helm because they are outside minibuffer.
> The help echo is displaying only mouse bindings, for all bindings you have a section with Commands in the org buffer when you hit C-h m.
> 
Ah… I took a small break and I was just about to go looking at the source to see.  I’ll check it out.
